router/core/graphql_prehandler.go

```diff
@@ -260,340 +273,304 @@ func (h *PreHandler) Handler(next http.Handler) http.Handler {
-			_, engineParseSpan := h.tracer.Start(r.Context(), "Operation - Parse",
-				trace.WithSpanKind(trace.SpanKindInternal),
-				trace.WithAttributes(commonAttributes...),
-				trace.WithAttributes(attributes...),
-			)
+		art.SetRequestTracingStats(r.Context(), traceOptions, traceTimings)
 
-			traceTimings.StartParse()
+		requestContext := buildRequestContext(w, r, opContext, requestLogger)
+		metrics.AddOperationContext(opContext)
 
-			err = operationKit.Parse()
-			if err != nil {
-				// the kit must be freed before we're doing io operations
-				// the kit is bound to the number of CPUs, and we must not hold onto it while doing IO operations
-				operationKit.Free()
-				operationKit = nil
-				finalErr = err
+		ctxWithRequest := withRequestContext(r.Context(), requestContext)
+		ctxWithOperation := withOperationContext(ctxWithRequest, opContext)
+		ww := middleware.NewWrapResponseWriter(w, r.ProtoMajor)
 
-				rtrace.AttachErrToSpan(engineParseSpan, err)
-				// Mark the root span of the router as failed, so we can easily identify failed requests
-				rtrace.AttachErrToSpan(routerSpan, err)
+		newReq := r.WithContext(ctxWithOperation)
 
-				engineParseSpan.End()
+		// Call the final handler that resolves the operation
+		// and enrich the context to make it available in the request context as well for metrics etc.
+		next.ServeHTTP(ww, newReq)
 
-				writeOperationError(r, w, requestLogger, err)
-				h.releaseBodyReadBuffer(buf)
-				return
-			}
+		statusCode = ww.Status()
+		writtenBytes = ww.BytesWritten()
 
-			traceTimings.EndParse()
-			engineParseSpan.End()
+		// Evaluate the request after the request has been handled by the engine handler
+		finalErr = requestContext.error
+
+		// Mark the root span of the router as failed, so we can easily identify failed requests
+		if finalErr != nil {
+			rtrace.AttachErrToSpan(trace.SpanFromContext(r.Context()), finalErr)
 		}
+	})
+}
 
-		h.releaseBodyReadBuffer(buf)
+func (h *PreHandler) handleOperation(req *http.Request, buf *bytes.Buffer, httpOperation *httpOperation) (*operationContext, error) {
 
-		// Set the router span name after we have the operation name
-		routerSpan.SetName(GetSpanName(operationKit.parsedOperation.Request.OperationName, operationKit.parsedOperation.Type))
+	operationKit, err := h.operationProcessor.NewKit(httpOperation.body, httpOperation.files)
+	defer func() {
+		// the kit must be freed before we're doing io operations
+		// the kit is bound to the number of CPUs, and we must not hold onto it while doing IO operations
+		// it needs to be called inside a defer to ensure it is called in panic situations as well
 
-		if blockedErr := h.operationBlocker.OperationIsBlocked(operationKit.parsedOperation); blockedErr != nil {
-			// the kit must be freed before we're doing io operations
-			// the kit is bound to the number of CPUs, and we must not hold onto it while doing IO operations
+		if operationKit != nil {
 			operationKit.Free()
 			operationKit = nil
```

Comment:
is this ineffectful?

Reply:
I was also under the same impression. I just copied the code over. I'm going to remove it.